File: python/trie.py

class TrieNode:
	#Trie node data structure

	def __init__(self):
		self.prefix=0
		self.word=0
		self.paths={}



class Trie:
	#Trie Tree

	def __init__(self):
		self.head=TrieNode()

	def add(self, word):
		#add a word to the Trie 

		word=word.lower()
		self.addToTrie(word, self.head)


	def addToTrie(self, word, node):
		if len(word) <= 0:
			node.word += 1
		else:
			node.prefix += 1
			firstChar=word[0]
			restWord=word[1:]

			if firstChar in node.paths:
				nextNode=node.paths[firstChar]
			else:
				nextNode=TrieNode()
				node.paths[firstChar]=nextNode

			self.addToTrie(restWord, nextNode)

	def isWord(self, word):
		#checks if the word is in the trie
		
		word=word.lower()
		numWords=self.wordCount(word, self.head)

		if numWords > 0:
			return True
		else:
			return False

	def wordCount(self, word, node):
		if len(word) <= 0:
			return node.word
		else:
			firstChar=word[0]
			restWord=word[1:]
				
			if firstChar not in node.paths:
				return 0
				
			nextNode=node.paths[firstChar]
			return self.wordCount(restWord, nextNode)

	def isPrefix(self, word):
		#checks if the word is a prefix to another word

		word=word.lower()
		numPrefix=self.prefixCount(word, self.head)

		if numPrefix > 0:
			return True
		else:
			return False

	def prefixCount(self, word, node):
		if len(word) <= 0:
			return node.prefix
		else:
			firstChar=word[0]
			restWord=word[1:]

			if firstChar not in node.paths:
				return 0

			nextNode=node.paths[firstChar]
			return self.prefixCount(restWord, nextNode)

File: Nodejs/trie.js
function TrieNode(){
	this.prefix=0;
	this.word=0;
	this.children=[];
}

TrieNode.prototype.addNode=function(character){
	if(!this.hasChar(character)){
		var newNode = new TrieNode();
		this.children[character] = newNode;
	}
}

TrieNode.prototype.hasChar = function(character){
	return character in this.children;
}

TrieNode.prototype.getCharNode = function(character){
	if(this.hasChar(character)){
		return this.children[character];
	}
	return undefined;
}




function Trie(){
	this.head=new TrieNode();
}



Trie.prototype.addWord = function(word){
		word=word.toLowerCase();
		this.addWordToTrie(word, this.head);
}

Trie.prototype.addWordToTrie = function(word, node){
	if(word.length <= 0 ){
		node.word = node.word + 1;
	}else{
		node.prefix = node.prefix + 1;
		var firstChar = word.substr(0,1);
		var restOfWord = word.substr(1);

		if(!node.hasChar(firstChar)){
			node.addNode(firstChar);
		}

		var nextNode = node.getCharNode(firstChar);
		this.addWordToTrie(restOfWord, nextNode);
	}
}

Trie.prototype.isWord = function(word){
	word=word.toLowerCase();
	var wordCount = this.wordCount(word, this.head);

	if(wordCount > 0){
		return true;
	}else{
		return false;
	}
}

Trie.prototype.wordCount = function(word, node){
	if(word.length <= 0){
		return node.word;
	}else{
		var firstChar = word.substr(0,1);
		var restOfWord = word.substr(1);

		if(!node.hasChar(firstChar)){
			return 0;
		}else{
			var nextNode = node.getCharNode(firstChar);
			return this.wordCount(restOfWord, nextNode);
		}
	}
}

Trie.prototype.isPrefix = function(word){
	word = word.toLowerCase();
	var prefixCount = this.prefixCount(word, this.head);

	if(prefixCount > 0){
		return true;
	}else{
		return false;
	}
}

Trie.prototype.prefixCount = function(word, node){
	if(word.length <= 0){
		return node.prefix;
	}else{
		var firstChar = word.substr(0,1);
		var restOfWord = word.substr(1);

		if(!node.hasChar(firstChar)){
			return 0;
		}else{
			var nextNode = node.getCharNode(firstChar);
			
			return this.prefixCount(restOfWord, nextNode);
		}
	}

}

module.exports.Trie=Trie;






